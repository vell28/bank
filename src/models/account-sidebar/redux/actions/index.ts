import { getStateType, IAction } from 'modules/store/types';
import { Dispatch } from 'redux';
import { push } from 'connected-react-router';

import { BANKING_PATH } from 'containers/routing/utils';
import { getRouterPathname } from 'models/router/selectors';
import { IAccountSidebar } from '../../entities';

export const UPDATE_ACCOUNT_SIDEBAR = 'sidebar/UPDATE_ACCOUNT';
export const SHOW_HIDE_SIDEBAR = 'sidebar/TOGGLE';

export type IUpdateSidebarAction = IAction<IAccountSidebar>;
export type IToggleSidebar = IAction<boolean>;

const BANKING_ACCOUNT = '/banking/account/';

export const updateSidebar = (data: IAccountSidebar): IUpdateSidebarAction => ({
  type: UPDATE_ACCOUNT_SIDEBAR,
  payload: data,
});

export const getSidebarAccountId = (data: IAccountSidebar) => (dispatch: Dispatch, setState: getStateType): any => {
  const pathname = getRouterPathname(setState());

  const path = pathname.replace(BANKING_ACCOUNT, '').replace('/', '');

  if (path && !data[path] && data[path] !== undefined) {
    dispatch(push(BANKING_PATH));
  }

  dispatch(updateSidebar(data));
};

export const toggleSidebar = (isShown: boolean): IToggleSidebar => ({
  type: SHOW_HIDE_SIDEBAR,
  payload: isShown,
});
