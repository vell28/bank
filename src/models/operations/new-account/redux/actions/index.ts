import {
  CurrencyCodeType,
} from 'modules/currencies';
import {
  IAction,
} from 'modules/store/types';
import { AccountTypeType } from '../../entities';

export const CREATE_ACCOUNT = 'accounts/CREATE_ACCOUNT';

export type ICreateAccountAction = IAction<{
  account: IAccountData;
}>

export interface IAccountData {
  accountType: AccountTypeType;
  accountCurrency: CurrencyCodeType;
}

export const createAccount = (
  account: IAccountData,
): ICreateAccountAction => ({
  type: CREATE_ACCOUNT,
  payload: { account },
});
