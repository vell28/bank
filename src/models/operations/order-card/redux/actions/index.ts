import { IAction, getStateType } from 'modules/store/types';
import { ICardOrderOwner, IAddress } from '../../entities';
import { CardProductType } from '../../../card-settings/entities';
import { ThunkAction } from './newOrder';
import { selectClientInfo, selectClientAddress } from '../selectors';

export const SET_CARD_PRODUCT_TYPE = 'orderNew/PRODUCT_TYPE';
export const SET_CARD_OWNER = 'orderNew/CARD_OWNER';
export const SET_CARD_DELIVERY_ADDRESS = 'orderNew/DELIVERY_ADDRESS';

export type ISetProductTypeAction = IAction<CardProductType | undefined>;

export const setCardProductType = (type?: CardProductType): ISetProductTypeAction => ({
  type: SET_CARD_PRODUCT_TYPE,
  payload: type,
});

export type ISetOwnerAction = IAction<ICardOrderOwner | undefined>;

export const setCardOwnner = (owner: ICardOrderOwner | undefined): ISetOwnerAction => ({
  type: SET_CARD_OWNER,
  payload: owner,
});

export type ISetDeliveryAddressAction = IAction<IAddress | undefined>;

export const setDeliveryAddress = (address?: IAddress): ISetDeliveryAddressAction => ({
  type: SET_CARD_DELIVERY_ADDRESS,
  payload: address,
});

export const orderCardSetDefaults: ThunkAction = (dispatch: any, getState: getStateType) => {
  const applicationState = getState();
  const ownerInfoByDefault = selectClientInfo(applicationState);
  const ownerAddressByDefault = selectClientAddress(applicationState);
  dispatch(setCardOwnner(ownerInfoByDefault));
  dispatch(setDeliveryAddress(ownerAddressByDefault));
};

export const orderCardCleanupState: ThunkAction = (dispatch: any) => {
  dispatch(setCardOwnner(undefined));
  dispatch(setDeliveryAddress(undefined));
};
