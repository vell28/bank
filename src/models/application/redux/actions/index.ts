import { fetchOrganizations } from 'models/organizations/redux/actions';
import { resetTransferLoadingState } from 'models/operations/transfer/redux/actions';
import { fetchTransactionsPage } from 'models/transactions/redux/actions';
import { isNotEmpty } from 'utils/ramda';

import { getOrganizations, getCurrentAccount, getClient } from 'models/organizations/redux/selectors';
import { getStateType } from 'modules/store/types';
import { ThunkAction } from 'utils/utilTypings';
import { saveAccountDetails, copyToAccountDetails } from 'modules/pdf/templates/accountDetails';
import { propOr } from 'ramda';
import { getBankName, getBankBIС } from '../..';

export const INIT_APP = 'application/SET';
export const UPDATE_APP = 'application/UPDATE';

export const initApp = async (dispatch: any, getState: getStateType) => {
  try {
    const organizations = getOrganizations(getState());
    // if organizations exist, fetch history and update organizations in backgound
    if (isNotEmpty(organizations)) {
      dispatch(fetchOrganizations());
      dispatch(fetchTransactionsPage());
    } else {
      await dispatch(fetchOrganizations());
      dispatch(fetchTransactionsPage());
    }
  } catch (e) {
    throw e;
  }
  dispatch(resetTransferLoadingState);
};

export const exportAccountDetails = (): ThunkAction => {
  return (dispatch: any, getState: getStateType) => {
    const state = getState();
    const account = getCurrentAccount(state);
    const bankTitle = getBankName(state);
    const bankTin = getBankBIС(state);
    const client = getClient(state);

    const iban: string = propOr('', 'id')(account);
    const recipientName: string = propOr('', 'name')(client);

    saveAccountDetails(iban, recipientName, bankTitle, bankTin);
  };
};

export const copyAccountDetails = (): ThunkAction => {
  return (dispatch: any, getState: getStateType) => {
    const state = getState();
    const account = getCurrentAccount(state);
    const bankTitle = getBankName(state);
    const bankTin = getBankBIС(state);
    const client = getClient(state);

    const iban: string = propOr('', 'id')(account);
    const recipientName: string = propOr('', 'name')(client);

    copyToAccountDetails(iban, recipientName, bankTitle, bankTin);
  };
};
