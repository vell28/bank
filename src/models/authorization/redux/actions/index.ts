import { Dispatch } from 'redux';
import { push } from 'connected-react-router';
import { find, equals } from 'ramda';
import { addSeconds } from 'date-fns';
import { SubmissionError } from 'redux-form';

import {
  passwordCheckRequest,
  logoutRequest,
  passwordCodeRequest,
  signInRequest,
  refreshTokenRequest,
} from 'modules/api-requests/authorization';
import { history } from 'modules/store';
import { smsConfirmFailedAttempt, smsConfirmClearAttempt } from 'models/operations/sms-confirmation/redux/actions';
import { PATHS_CONFIG, AUTHORIZATION_PATH } from 'containers/routing/utils';
import { firebaseAuth } from 'modules/firebase-auth';
import { getStateType } from 'modules/store/types';
import { IServerErrors } from 'modules/api-client/entities';

import { EXPIRES_FIREBASE_SECOND_INTERVAL, EXPIRES_UAS_SECOND_INTERVAL } from 'modules/api-client/utils/const';
import { IPasswordCheckRequest, ICodeRequest } from '../../entities';
import { getSessId, getPhone, getFirebaseToken } from '../selectors';
import { DEV_TEST_PHONE_LIST, AUTH_CONFIRM_ERROR, AUTH_PASSWORD_ERROR } from './const';
// import { isDevEnv } from 'utils/enviroment';

export const AUTH_SET_TOKEN = 'authorization/SET_TOKEN';
export const AUTH_UPDATE_TOKEN = 'authorization/UPDATE_TOKEN';
export const AUTH_REMOVE_TOKEN = 'authorization/REMOVE_TOKEN';
export const AUTH_CHECK_PASSWORD = 'authorization/PASSWORD_CHECK';
export const AUTH_REQUEST_CODE = 'authorization/REQUEST_CODE';
export const TOGGLE_LOGOUT_CONFIRM_MODAL = 'authorization/LOGOUT_CONFIRM_MODAL';

export interface IUpdateTokenAction {
  type: string;
  payload: {
    token: string;
    token_expires: Date;
  };
}

// TODO: in prod and stage use import { isDevEnv } from 'utils/enviroment';
const isDevEnv = true;

export const refreshToken = (token: string, expires: Date): IUpdateTokenAction => ({
  type: AUTH_UPDATE_TOKEN,
  payload: {
    token,
    token_expires: expires,
  },
});

export const authRequestCodeAction = (dispatch: Dispatch) =>
  dispatch({
    type: AUTH_REQUEST_CODE,
    payload: async () => {
      const result = await passwordCodeRequest();
      if (result.ok) {
        dispatch(push(PATHS_CONFIG.signin_confirm.path));
        return result.data;
      }

      throw result;
    },
  });

export const passwordCheckAction = (formData: IPasswordCheckRequest) => (dispatch: Dispatch) =>
  dispatch({
    type: AUTH_CHECK_PASSWORD,
    payload: async () => {
      const { phone } = formData;
      try {
        const isTestPhone = !!find(equals(phone))(DEV_TEST_PHONE_LIST);
        const result = await passwordCheckRequest(formData);
        if (result.ok && result.data.status === 'ok') {
          if (isDevEnv && isTestPhone) {
            setTimeout(() => authRequestCodeAction(dispatch));
          } else {
            await firebaseAuth.getLoginCode(phone);
          }
          dispatch(smsConfirmClearAttempt);
        }
        dispatch(push(PATHS_CONFIG.signin_confirm.path));
        return { phone };
      } catch (e) {
        throw new SubmissionError({ password: AUTH_PASSWORD_ERROR });
      }
    },
  });

export const signIn = (data: ICodeRequest) => (dispatch: Dispatch, getState: getStateType) =>
  dispatch({
    type: AUTH_SET_TOKEN,
    payload: async () => {
      const sessid = getSessId(getState());
      const phone = getPhone(getState());

      try {
        const isTestPhone = !!find(equals(phone))(DEV_TEST_PHONE_LIST);
        if (isTestPhone && isDevEnv) {
          const result = await signInRequest({ ...data, sessid });
          // TODO: fix it, replace to single action
          setTimeout(() => dispatch(push(PATHS_CONFIG.banking.path)), 0);
          return result;
        }

        const { code } = data;
        if (code) {
          const tokenFirebase = await firebaseAuth.verifyCode(code);
          setTimeout(() => dispatch(push(PATHS_CONFIG.banking.path)), 0);
          return { 'token-firebase': tokenFirebase };
        }
      } catch (e) {
        dispatch(smsConfirmFailedAttempt);
        throw new SubmissionError({ code: AUTH_CONFIRM_ERROR });
      }
    },
  });

export const refreshTokenAction = async (dispatch: Dispatch, getState: getStateType) => {
  const firebaseToken = getFirebaseToken(getState());
  if (firebaseToken) {
    const token = await firebaseAuth.refreshToken();
    const expiredTime = addSeconds(Date.now(), EXPIRES_FIREBASE_SECOND_INTERVAL);
    return dispatch(refreshToken(token, expiredTime));
  }
  const response = await refreshTokenRequest();
  if (response.ok) {
    return dispatch(refreshToken(response.data.token, addSeconds(Date.now(), EXPIRES_UAS_SECOND_INTERVAL)));
  }
};

export const setTokenError = (errors: IServerErrors) => ({
  type: `${AUTH_UPDATE_TOKEN}_ERROR`,
  payload: { errors },
});

export const removeToken = {
  type: AUTH_REMOVE_TOKEN,
};

export const toggleLogoutModal = (isShown: boolean) => ({
  type: TOGGLE_LOGOUT_CONFIRM_MODAL,
  payload: isShown,
});

export const goToAuth = (dispatch: Dispatch) => {
  dispatch(removeToken);
  dispatch(toggleLogoutModal(false));
  dispatch(push(AUTHORIZATION_PATH));
};

export const logout = async (dispatch: Dispatch) => {
  try {
    const result = await logoutRequest();
    if (result.ok && result.data.success) {
      dispatch(removeToken);
      history.push(PATHS_CONFIG.login.path);
    }
  } catch (e) {
    return e;
  }
};
