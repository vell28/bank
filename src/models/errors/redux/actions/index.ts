import { Dispatch } from 'redux';
import { push } from 'connected-react-router';

import { ERROR_PATH } from 'containers/routing/utils';
import { IAction } from 'modules/store/types';
import {
  default404, default500, default403, defaultInternet, expiredSession
} from './const';

import { IAppErrors } from '../../entities';

export const SET_APP_ERROR = 'app-error/SET';

export type IAppErrorAction = IAction<IAppErrors>;

export const app404Error = (dispatch: Dispatch) => (error?: IAppErrors) => {
  dispatch({
    type: SET_APP_ERROR,
    payload: default404(error),
  });
  dispatch(push(ERROR_PATH));
};

export const app500Error = (dispatch: Dispatch) => (error?: IAppErrors) => {
  dispatch({
    type: SET_APP_ERROR,
    payload: default500(error),
  });
  dispatch(push(ERROR_PATH));
};

export const app403Error = (dispatch: Dispatch) => (error?: IAppErrors) => {
  dispatch({
    type: SET_APP_ERROR,
    payload: default403(error),
  });
  dispatch(push(ERROR_PATH));
};

export const appInternetError = (dispatch: Dispatch) => (error?: IAppErrors) => {
  dispatch({
    type: SET_APP_ERROR,
    payload: defaultInternet(error),
  });
  dispatch(push(ERROR_PATH));
};

export const appSessionExpired = (dispatch: Dispatch) => {
  dispatch({
    type: SET_APP_ERROR,
    payload: expiredSession,
  });
  dispatch(push(ERROR_PATH));
};
