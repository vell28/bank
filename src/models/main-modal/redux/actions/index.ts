import { Dispatch } from 'redux';

import { getCurrentRouterPath } from 'models/application/redux/selectors';

import { IAction, getStateType } from 'modules/store/types';
import { getTopUpIsError } from 'models/operations/top-up/redux/selectors';
import { getTransferIsError } from 'models/operations/transfer/redux/selectors';
import { updateTopUpResponse } from 'models/operations/top-up/redux/actions';
import { updateTransferResponse } from 'models/operations/transfer/redux/actions';
import { ERROR_MODAL_TITLE } from 'containers/operations/transaction-error/constant';
import { getMainModalContentId } from '../selectors';
import { IMainModal, MainModalContentType, IMainModalSettings } from '../../entities';

export const TRANSACTION_ERROR_MODAL = 'main-modal/TRANSACTION_ERROR_MODAL';
export const UPDATE_MAIN_MODAL = 'main-modal/UPDATE';
export const TOGGLE_MAIN_MODAL = 'main-modal/TOGGLE';
export const SHOW_WITH_CONTENT_MAIN_MODAL = 'main-modal/SHOW_WITH_CONTENT';
export const NEXT_STEP_MAIN_MODAL = 'main-modal/NEXT_STEP';
export const TARGET_STEP_MAIN_MODAL = 'main-modal/TARGET_STEP';

export type IUpdateMainModalAction = IAction<IMainModal>;
export type IToggleMainModalAction = IAction<null>;
export type IShowContentMainModalAction = IAction<IMainModalSettings>;
export type ITargetStepMainModalAction = IAction<number>;
export interface IErrorClear {
  isError: boolean;
  errors: any[];
}

export const updateMainModal = (data: IMainModal): IUpdateMainModalAction => ({
  type: UPDATE_MAIN_MODAL,
  payload: data,
});

export const toggleMainModal: IToggleMainModalAction = {
  type: TOGGLE_MAIN_MODAL,
  payload: null,
};

export const showModalWithContent = (contentId: MainModalContentType, title?: string) => (
  dispatch: Dispatch,
  getState: getStateType,
): IShowContentMainModalAction => {
  const bindUrl = getCurrentRouterPath(getState());
  return dispatch({
    type: SHOW_WITH_CONTENT_MAIN_MODAL,
    payload: { id: contentId, bindUrl, title },
  });
};

export const showTransactionErrorModal = () => (dispatch: Dispatch, getState: getStateType) => {
  const topUpError = getTopUpIsError(getState());
  const transferError = getTransferIsError(getState());
  const id = getMainModalContentId(getState());
  const payload = { isError: false, errors: [] };

  if (transferError) {
    dispatch(updateTransferResponse({ payload }));
  }

  if (topUpError) {
    dispatch(updateTopUpResponse({ payload }));
  }

  const title = ERROR_MODAL_TITLE[id];

  dispatch<any>(showModalWithContent(TRANSACTION_ERROR_MODAL, title));
};

export const nextStep = {
  type: NEXT_STEP_MAIN_MODAL,
  payload: null,
};

export const targetStep = (step: number): ITargetStepMainModalAction => ({
  type: TARGET_STEP_MAIN_MODAL,
  payload: step,
});
