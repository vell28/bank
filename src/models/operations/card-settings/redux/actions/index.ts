import { Dispatch } from 'redux';

import {
  fetchUnmaskedCardRequest,
  setCardStatusRequest,
  setUnBlockCardRequest,
  ICardStatusData,
  setCardLimitRequest,
  setDisableCardLimitsRequest,
  setIncreaseCardLimitRequest,
} from 'modules/api-requests/cards';

import { AccountStatus } from 'models/organizations/entities';
import { getCurrentAccountId } from 'models/organizations/redux/selectors';
import { getConfirmCodeLengthToken, getConfirmFromResponseToken } from 'modules/api-client/utils';
import { isNotEmpty } from 'utils/ramda';

import { getStateType, IAction, IAsyncAction } from 'modules/store/types';
import { ICardUnmasked, ICard, ILimitData } from '../../entities';
import { getIncreaseLimit } from '../../../card-settings/redux/selectors';
import { getConfirmationToken } from '../../../sms-confirmation/redux/selectors';
import {
  setSmsConfirmData,
  smsConfirmClearAttempt,
  smsConfirmFailedAttempt,
} from '../../../sms-confirmation/redux/actions';

export const MAX_FAILED_ATTEMPTS = 3;

export const FETCH_CARD_UNMASKED = 'cards/FETCH_UNMASKED';
export const LOCK_CARD = 'cards/LOCK';
export const UNLOCK_CARD = 'cards/UN_LOCK';
export const UNLOCK_CARD_CONFIRM = 'cards/UNLOCK_CARD_CONFIRM';
export const SET_SELECTED_CARD = 'cards/SELECT_CARD_IN_SLIDER';

export const TOGGLE_SHOWN_BLOCK_CARD = 'cards/TOGGLE_SHOWN_BLOCK';
export const TOGGLE_SHOWN_UNBLOCK_CARD = 'cards/TOGGLE_SHOWN_UNBLOCK';
export const TOGGLE_SHOWN_UNBLOCK_CONFIRM_CARD = 'cards/TOGGLE_SHOWN_CONFIRM_UNBLOCK';
export const REDUCTION_CARD_LIMIT = 'cards/REDUCTION_LIMIT';
export const SET_INCREASED_CARD_LIMIT = 'cards/SET_INCREASED_LIMIT';
export const INCREASE_CARD_LIMIT = 'cards/INCREASE_LIMIT';
export const INCREASE_CARD_LIMIT_CONFIRM = 'cards/INCREASE_LIMIT_CONFIRM';
export const DISABLE_CARD_LIMITS = 'cards/DISABLE_LIMITS';
export const DISABLE_CARD_LIMITS_CONFIRM = 'cards/DISABLE_LIMITS_CONFIRM';
export const TOGGLE_SHOWN_DISABLE_CARD_LIMITS = 'cards/TOGGLE_SHOWN_DISABLE_LIMITS';
export const TOGGLE_SHOWN_DISABLE_CARD_LIMITS_CONFIRM = 'cards/TOGGLE_SHOWN_DISABLE_LIMITS_CONFIRM';
export const TOGGLE_SHOWN_CARD_LIMITS_TIMER = 'cards/TOGGLE_SHOWN_LIMITS_TIMER';
export const TOGGLE_SHOWN_INCREASE_CARD_LIMIT = 'cards/TOGGLE_SHOWN_INCREASE_LIMIT';
export const TOGGLE_SHOWN_INCREASE_CARD_LIMIT_CONFIRM = 'cards/TOGGLE_SHOWN_INCREASE_LIMIT_CONFIRM';

export interface ISelectedCard {
  card: ICard;
  index: number;
}

export type IToggleCardModal = IAction<boolean>;

export type ISelectCardAction = IAction<{
  [key: string]: ISelectedCard;
}>;

export interface IFetchUnmaskedAction extends IAsyncAction<ICardUnmasked> {
  meta: string;
}

export interface ILockCardAction extends IAsyncAction<ICard> {
  meta: string;
}

export interface IIncreaseCardLimitAction extends IAsyncAction<ICard> {
  meta: string;
}

export type ISetIncreasedLimitsAction = IAction<ILimitData>;

export const fetchUnmaskedCard = (cardId: string): IFetchUnmaskedAction => ({
  type: FETCH_CARD_UNMASKED,
  payload: fetchUnmaskedCardRequest(cardId),
  meta: cardId,
});

export const toggleUnBlockCardModal = (isShown: boolean): IToggleCardModal => ({
  type: TOGGLE_SHOWN_UNBLOCK_CARD,
  payload: isShown,
});

export const toggleUnBlockConfirmCardModal = (isShown: boolean): IToggleCardModal => ({
  type: TOGGLE_SHOWN_UNBLOCK_CONFIRM_CARD,
  payload: isShown,
});

export const lockCard = (cardId: string, accountId: string): ILockCardAction => {
  const requestData: ICardStatusData = {
    data: {
      card: { status: AccountStatus.LOCKED },
    },
  };

  return {
    type: LOCK_CARD,
    payload: setCardStatusRequest(cardId, requestData),
    meta: accountId,
  };
};

export const unlockCard = (cardId: string) => (dispatch: Dispatch) => {
  return dispatch({
    type: UNLOCK_CARD,
    payload: async () => {
      try {
        dispatch(smsConfirmClearAttempt);
        return await setUnBlockCardRequest(cardId, AccountStatus.ACTIVE);
      } catch (e) {
        const confirmationToken = getConfirmFromResponseToken(e);
        if (isNotEmpty(confirmationToken)) {
          const confirmationCodeLength = getConfirmCodeLengthToken(e);
          dispatch(setSmsConfirmData({ confirmationToken, confirmationCodeLength }));
          dispatch(toggleUnBlockCardModal(false));
          dispatch(toggleUnBlockConfirmCardModal(true));
        } else {
          throw e;
        }
      }
    },
  });
};

export const unblockCardConfirm = (cardId: string, accountId: string, code: string) => (
  dispatch: Dispatch,
  getState: getStateType,
) => {
  const state = getState();
  const token = getConfirmationToken(state);
  return dispatch({
    type: UNLOCK_CARD_CONFIRM,
    payload: async () => {
      try {
        const result = await setUnBlockCardRequest(cardId, AccountStatus.ACTIVE, {
          'X-Confirmation-Code': code,
          'X-Confirmation-Token': token,
        });
        dispatch(smsConfirmClearAttempt);
        return result;
      } catch (e) {
        dispatch(smsConfirmFailedAttempt);
        throw e;
      }
    },
    meta: accountId,
  });
};

export const selectCardInSlider = (accountId: string, card: ICard, index: number): ISelectCardAction => ({
  type: SET_SELECTED_CARD,
  payload: { [accountId]: { card, index } },
});

export const toggleBlockCardModal = (isShown: boolean): IToggleCardModal => ({
  type: TOGGLE_SHOWN_BLOCK_CARD,
  payload: isShown,
});

export const toggleDisableCardLimitsModal = (isShown: boolean): IToggleCardModal => ({
  type: TOGGLE_SHOWN_DISABLE_CARD_LIMITS,
  payload: isShown,
});

export const toggleDisableCardLimitsConfirmModal = (isShown: boolean): IToggleCardModal => ({
  type: TOGGLE_SHOWN_DISABLE_CARD_LIMITS_CONFIRM,
  payload: isShown,
});

export const toggleDisableCardLimitsTimer = (isShown: boolean): IToggleCardModal => ({
  type: TOGGLE_SHOWN_CARD_LIMITS_TIMER,
  payload: isShown,
});

export const toggleIncreaseCardLimitModal = (isShown: boolean): IToggleCardModal => ({
  type: TOGGLE_SHOWN_INCREASE_CARD_LIMIT,
  payload: isShown,
});

export const toggleIncreaseCardLimitConfirmModal = (isShown: boolean): IToggleCardModal => ({
  type: TOGGLE_SHOWN_INCREASE_CARD_LIMIT_CONFIRM,
  payload: isShown,
});

export const reductionLimit = (cardId: string, limit: ILimitData) => (dispatch: Dispatch, getState: getStateType) => {
  const state = getState();
  dispatch({
    type: REDUCTION_CARD_LIMIT,
    payload: setCardLimitRequest(cardId, limit),
    meta: getCurrentAccountId(state),
  });
};

export const setIncreasedCardLimit = (limit: ILimitData): ISetIncreasedLimitsAction => ({
  type: SET_INCREASED_CARD_LIMIT,
  payload: limit,
});

export const increaseLimit = (cardId: string, limit: ILimitData) => (dispatch: Dispatch) => {
  dispatch(toggleIncreaseCardLimitModal(true));
  dispatch(setIncreasedCardLimit(limit));
  dispatch({
    type: INCREASE_CARD_LIMIT,
    payload: async () => {
      try {
        const result = await setIncreaseCardLimitRequest(cardId, limit);
        dispatch(smsConfirmClearAttempt);
        return result;
      } catch (e) {
        const confirmationToken = getConfirmFromResponseToken(e);
        if (isNotEmpty(confirmationToken)) {
          const confirmationCodeLength = getConfirmCodeLengthToken(e);
          dispatch(setSmsConfirmData({ confirmationToken, confirmationCodeLength }));
          dispatch(toggleIncreaseCardLimitModal(false));
          dispatch(toggleIncreaseCardLimitConfirmModal(true));
        } else {
          throw e;
        }
      }
    },
  });
};

export const increaseLimitConfirm = (cardId: string, code: string) => (dispatch: Dispatch, getState: getStateType) => {
  const state = getState();
  const token = getConfirmationToken(state);
  const limit = getIncreaseLimit(state);
  return dispatch({
    type: INCREASE_CARD_LIMIT_CONFIRM,
    payload: async () => {
      try {
        const result = await setIncreaseCardLimitRequest(cardId, limit, {
          'X-Confirmation-Code': code,
          'X-Confirmation-Token': token,
        });
        dispatch(smsConfirmClearAttempt);
        return result;
      } catch (e) {
        dispatch(smsConfirmFailedAttempt);
        throw e;
      }
    },
    meta: getCurrentAccountId(state),
  });
};

export const disableCardLimits = (cardId: string) => (dispatch: Dispatch) => {
  return dispatch({
    type: DISABLE_CARD_LIMITS,
    payload: async () => {
      try {
        dispatch(smsConfirmClearAttempt);
        return await setDisableCardLimitsRequest(cardId);
      } catch (e) {
        const confirmationToken = getConfirmFromResponseToken(e);
        if (isNotEmpty(confirmationToken)) {
          const confirmationCodeLength = getConfirmCodeLengthToken(e);
          dispatch(setSmsConfirmData({ confirmationToken, confirmationCodeLength }));
          dispatch(toggleDisableCardLimitsModal(false));
          dispatch(toggleDisableCardLimitsConfirmModal(true));
        } else {
          throw e;
        }
      }
    },
  });
};

export const disableCardLimitsConfirm = (cardId: string, code: string) => (
  dispatch: Dispatch,
  getState: getStateType,
) => {
  const state = getState();
  const token = getConfirmationToken(state);
  return dispatch({
    type: DISABLE_CARD_LIMITS_CONFIRM,
    payload: async () => {
      try {
        const result = await setDisableCardLimitsRequest(cardId, {
          'X-Confirmation-Code': code,
          'X-Confirmation-Token': token,
        });
        dispatch(smsConfirmClearAttempt);
        return result;
      } catch (e) {
        dispatch(smsConfirmFailedAttempt);
        throw e;
      }
    },
    meta: getCurrentAccountId(state),
  });
};
