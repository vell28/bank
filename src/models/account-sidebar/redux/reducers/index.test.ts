import { accountSidebar, IAccountSidebarState } from '.';
import { UPDATE_ACCOUNT_SIDEBAR, SHOW_HIDE_SIDEBAR } from '../actions';

const initAccountSidebar: IAccountSidebarState = {
  data: {},
  isShown: false,
};

const accountId = 'accountId';

const data = {
  [accountId]: true,
};

describe('accountSidebar reducers', () => {
  test('accountSidebar/UPDATE_ACCOUNT_SIDEBAR', () => {
    const action = {
      type: UPDATE_ACCOUNT_SIDEBAR,
      payload: data,
    };

    const newState = accountSidebar(initAccountSidebar, action);
    expect(newState.data[accountId]).toBeTruthy();
    expect(newState).toMatchSnapshot();
  });

  test('accountSidebar/UPDATE_ACCOUNT_SIDEBAR - should hide sidebar', () => {
    const action = {
      type: SHOW_HIDE_SIDEBAR,
      payload: false,
    };

    const newState = accountSidebar(initAccountSidebar, action);
    expect(newState.isShown).toBeFalsy();
    expect(newState).toMatchSnapshot();
  });

  test('accountSidebar/UPDATE_ACCOUNT_SIDEBAR - should show sidebar', () => {
    const action = {
      type: SHOW_HIDE_SIDEBAR,
      payload: true,
    };

    const newState = accountSidebar(initAccountSidebar, action);
    expect(newState.isShown).toBeTruthy();
    expect(newState).toMatchSnapshot();
  });
});
