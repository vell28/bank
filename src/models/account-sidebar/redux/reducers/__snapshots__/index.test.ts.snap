// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`accountSidebar reducers accountSidebar/UPDATE_ACCOUNT_SIDEBAR - should hide sidebar 1`] = `
Object {
  "data": Object {},
  "isShown": false,
}
`;

exports[`accountSidebar reducers accountSidebar/UPDATE_ACCOUNT_SIDEBAR - should show sidebar 1`] = `
Object {
  "data": Object {},
  "isShown": true,
}
`;

exports[`accountSidebar reducers accountSidebar/UPDATE_ACCOUNT_SIDEBAR 1`] = `
Object {
  "data": Object {
    "accountId": true,
  },
  "isShown": false,
}
`;
